pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+      setLastPurgeRunStatus(PurgeRunStatus.SUCCESS);
+    } catch (IOException e) {
+      setLastPurgeRunStatus(PurgeRunStatus.FAILURE);
+      logger.error("Error occurred while purging older snapshots", e);
+    }
+  }
+
+  private void deleteOlderSnapshots(
+      List<ResourceId> allPaths, TreeSet<String> recentSnapshotsToBeRetained) throws IOException {
+    for (ResourceId path : allPaths) {
+      if (!recentSnapshotsToBeRetained.contains(path.getFilename()) && isDWHComplete(path)) {
+        deleteDirectoryAndFiles(path);
+      }
+    }
+  }
+
+  /**
+   * This method deletes all the files, directories and subdirectories under the given root
+   * directory. This is done by first deleting all the files under the root directory including
+   * under the subdirectories. Later the directories and the subdirectories are deleted.
+   *
+   * @param rootDirectory which needs to be deleted
+   * @throws IOException
+   */
+  public void deleteDirectoryAndFiles(ResourceId rootDirectory) throws IOException {
+    ResourceId filesResourceToBeMatched =
+        rootDirectory.resolve("**", StandardResolveOptions.RESOLVE_FILE);
+    List<MatchResult> matchedFilesResultList =
+        FileSystems.matchResources(Collections.singletonList(filesResourceToBeMatched));
+
+    // Collect the matched files which also includes the files under the subdirectories, at the same
+    // time collect the directories as well.
+    List<ResourceId> fileResources = new ArrayList<>();
+    Set<String> allDirectoryPathSet = new HashSet<>();
+    for (MatchResult matchResult : matchedFilesResultList) {
+      if (matchResult.status() == Status.OK && !matchResult.metadata().isEmpty()) {
+        for (Metadata metadata : matchResult.metadata()) {
+          fileResources.add(metadata.resourceId());
+          ResourceId currentDirectory = metadata.resourceId().getCurrentDirectory();
+          allDirectoryPathSet.add(currentDirectory.toString());
+          // Beam currently creates a temporary empty folder with name '.temp-beam' sometimes under
+          // the directories. These empty subdirectories could not be determined with the existing
+          // Beam file system apis, hence this is hard-coded here so that these empty directories
+          // and their parent directories can be successfully deleted.
+          allDirectoryPathSet.add(
+              currentDirectory
+                  .resolve(".temp-beam", StandardResolveOptions.RESOLVE_DIRECTORY)
+                  .toString());
+        }
+      } else if (matchResult.status() == Status.ERROR) {
+        String errorMessage =
+            String.format("Error matching files under directory %s", rootDirectory.getFilename());
+        logger.error(errorMessage);
+        throw new IOException(errorMessage);
+      }
+    }
+    List<String> allDirectoryPathList = new ArrayList<>(allDirectoryPathSet);
+    allDirectoryPathList.sort(Collections.reverseOrder());
```

Comment:
Why do we need the sort here?

Reply:
Sorting is done so that, first the leaf child directories gets deleted, followed by their corresponding parent directories. Added a comment in the code explaining the same.

Reply:
SG; thanks.